sql/service_wsrep.cc

```diff
@@ -183,7 +183,21 @@ extern "C" my_bool wsrep_thd_is_BF(const THD *thd, my_bool sync)
 
 extern "C" my_bool wsrep_thd_is_SR(const THD *thd)
 {
-  return thd && thd->wsrep_cs().transaction().is_streaming();
+  return thd && thd->wsrep_cs().transaction().is_streaming() &&
+    thd->wsrep_cs().transaction().state() == wsrep::transaction::s_executing;
+}
+
+extern "C" my_bool wsrep_thd_is_XA_prepared(const MYSQL_THD thd)
```

Comment:
In my opinion, this part should be dropped. If you want to fix it for the XA case you need to grab a copy of mariadb enterprise, so that you can test it and fix it there.